core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/RackAwareWeightedLoadBalancingPolicy.java

```diff
@@ -0,0 +1,264 @@
+        }
+      }
+    }
+
+    // collect any non-replicas, if we need to and there are some available
+    if (nodeScores.size() < this.planSize && nodeScores.size() < dcNodeSet.size()) {
+      Random rand = getRandom();
+      final Node[] dcNodes = dcNodeSet.toArray(new Node[0]);
+
+      for (int i = 0; i < dcNodes.length; i++) {
+        // pick a random target; shuffle it up front, so we don't revisit
+        int nextIndex = i + rand.nextInt(dcNodes.length - i);
+        ArrayUtils.swap(dcNodes, i, nextIndex);
+        Node dcNode = dcNodes[i];
+
+        // skip replicas; they were already inserted
+        // otherwise, found a valid node: score it
+        if (!replicaSet.contains(dcNode)) {
+          nodeScores.add(new NodeScore(dcNode,
+                  getWeightedScore(dcNode, session, nowMillis, nowNanos, false)));
+
+          // if we scored, we might by now have already scored enough of what we need
+          if (nodeScores.size() == this.planSize || nodeScores.size() == dcNodes.length) {
+            break;
+          }
+        }
+      }
+
+      // by now we've scored everything we need to meet planSize, or if not, at least everything available
+    }
+
+    // At this point we have a small, typically 8 element array containing all local
+    // datacenter replicas and potentially some random choices from the rest of the datacenter.
+    //
+    // We now rank nodes by a score function that takes into account outstanding requests weighted
+    // by replica status, rack placement, uptime, and health status. In general, we expect to
+    // see the following order
+    //   1. Rack replicas
+    //   2. Datacenter replicas
+    //   3. Rack nodes
+    //   4. Datacenter nodes
+    // We expect these orderings to move around when nodes are overloaded. For example if the
+    // local zone replica has too much load we will failover to other replicas. If those
+    // replicas are too slow we will failover to other rack nodes.
+
+    // sort, extract, convert to query plan
+    nodeScores.sort(Comparator.comparingDouble(NodeScore::getScore));
+    Node[] candidate = new Node[nodeScores.size()];
+    for (int i = 0; i < candidate.length; i++) {
+      candidate[i] = nodeScores.get(i).getNode();
+    }
+
+    QueryPlan queryPlan = candidate.length == 0 ? QueryPlan.EMPTY : new SimpleQueryPlan((Object[]) candidate);
+    return maybeAddDcFailover(request, queryPlan);
+  }
+
+  protected String getLocalRack() {
+    return ""; // TODO (akhaku) internally we passed it through the context, for OSS perhaps something like the local DC helper?
+  }
+
+  protected boolean inRack(Node node) {
+    if (node == null || node.getRack() == null) return false;
+    return node.getRack().equals(this.getLocalRack());
+  }
+
+  protected double getWeightedScore(Node node, Session session, long nowMillis, long nowNanos, boolean isReplica) {
+    int base = Math.min(32768, 1 + getInFlight(node, session));
```

Comment:
`getInFlight` calls `InFlightHandler#getInFlight`, which isn't thread-safe. `StreamIdGenerator#maxAvailableIds` is a plain int, so could be cached on a calling thread. Risks of this value being inaccurate are low, could update `maxAvailableIds` to be a volatile, either way not a huge concern.

Reply:
`maxAvailableIds` is a `final` instance variable on `StreamIdGenerator` so I think we're safe here, unless I misunderstand how that works. And `availableIds()` returns the value of an AtomicInteger, so I think that means that `InFlightHandler#getInFlight` is thread-safe

Reply:
You're right - it's final, missed that.